src/common/io/http/response.c

```diff
@@ -0,0 +1,412 @@
+}
+
+/***********************************************************************************************************************************
+Has all content been read?
+***********************************************************************************************************************************/
+static bool
+httpResponseEof(THIS_VOID)
+{
+    THIS(HttpResponse);
+
+    FUNCTION_LOG_BEGIN(logLevelTrace);
+        FUNCTION_LOG_PARAM(HTTP_RESPONSE, this);
+    FUNCTION_LOG_END();
+
+    ASSERT(this != NULL);
+
+    FUNCTION_LOG_RETURN(BOOL, this->contentEof);
+}
+
+/**********************************************************************************************************************************/
+HttpResponse *
+httpResponseNew(HttpSession *session, const String *verb, bool contentCache)
+{
+    FUNCTION_LOG_BEGIN(logLevelDebug)
+        FUNCTION_LOG_PARAM(HTTP_SESSION, session);
+        FUNCTION_LOG_PARAM(STRING, verb);
+        FUNCTION_LOG_PARAM(BOOL, contentCache);
+    FUNCTION_LOG_END();
+
+    ASSERT(session != NULL);
+    ASSERT(verb != NULL);
+
+    HttpResponse *this = NULL;
+
+    MEM_CONTEXT_NEW_BEGIN("HttpResponse")
+    {
+        this = memNew(sizeof(HttpResponse));
+
+        *this = (HttpResponse)
+        {
+            .memContext = MEM_CONTEXT_NEW(),
+            .session = httpSessionMove(session, memContextCurrent()),
+        };
+
+        // Read status
+        String *status = ioReadLine(httpSessionIoRead(this->session));
+
+        // Check status ends with a CR and remove it to make error formatting easier and more accurate
+        if (!strEndsWith(status, CR_STR))
+            THROW_FMT(FormatError, "HTTP response status '%s' should be CR-terminated", strPtr(status));
+
+        status = strSubN(status, 0, strSize(status) - 1);
+
+        // Check status is at least the minimum required length to avoid harder to interpret errors later on
+        if (strSize(status) < sizeof(HTTP_VERSION) + 4)
+            THROW_FMT(FormatError, "HTTP response '%s' has invalid length", strPtr(strTrim(status)));
+
+        // Check status starts with the correct http version
+         if (!strBeginsWith(status, HTTP_VERSION_STR))
+            THROW_FMT(FormatError, "HTTP version of response '%s' must be " HTTP_VERSION, strPtr(status));
+
+        // Read status code
+        status = strSub(status, sizeof(HTTP_VERSION));
+
+        int spacePos = strChr(status, ' ');
+
+        if (spacePos != 3)
+            THROW_FMT(FormatError, "response status '%s' must have a space after the status code", strPtr(status));
+
+        this->code = cvtZToUInt(strPtr(strSubN(status, 0, (size_t)spacePos)));
+
+        // Read reason phrase. A missing reason phrase will be represented as an empty string.
+        MEM_CONTEXT_BEGIN(this->memContext)
+        {
+            this->reason = strSub(status, (size_t)spacePos + 1);
+        }
+        MEM_CONTEXT_END();
+
+        // Read headers
+        MEM_CONTEXT_BEGIN(this->memContext)
```

Comment:
Is there a reason when the "this->reason = " and the "this->header = " cannot be combined under the same MEM_CONTEXT_BEGIN(this->memContext)

Reply:
No, but actually the mem contexts were not quite right here anyway. This code was copied from httpClient where it lived inside a temp context, but here we are already in the `this->memContext` context. What we need is a temp context to wrap it. And creating the header list can going into the object's designated initializer.

So, did all that.

The way it is currently the temp strings created while parsing the response would have the same lifetime as the response object. Not a big deal, perhaps, but not as efficient as it could be.